consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinator.java

```diff
@@ -188,28 +197,86 @@ public PayloadIdentifier preparePayload(
     }
 
+    tryToBuildBetterBlock(timestamp, random, payloadIdentifier, mergeBlockCreator);
+
+    return payloadIdentifier;
+  }
+
+  private void tryToBuildBetterBlock(
+      final Long timestamp,
+      final Bytes32 random,
+      final PayloadIdentifier payloadIdentifier,
+      final MergeBlockCreator mergeBlockCreator) {
+
+    long remainingTime = miningParameters.getPosBlockCreationTimeout();
+    final Supplier<Block> blockCreator =
+        () -> mergeBlockCreator.createBlock(Optional.empty(), random, timestamp);
     // start working on a full block and update the payload value and candidate when it's ready
-    CompletableFuture.supplyAsync(
-            () -> mergeBlockCreator.createBlock(Optional.empty(), random, timestamp))
-        .orTimeout(12, TimeUnit.SECONDS)
-        .whenComplete(
-            (bestBlock, throwable) -> {
-              if (throwable != null) {
-                LOG.debug("something went wrong creating block", throwable);
+    final long startedAt = System.currentTimeMillis();
+    retryBlockCreation(payloadIdentifier, blockCreator, remainingTime)
+        .thenAccept(
+            bestBlock -> {
+              final var resultBest = validateBlock(bestBlock);
+              if (resultBest.blockProcessingOutputs.isPresent()) {
+                mergeContext.putPayloadById(payloadIdentifier, bestBlock);
+                debugLambda(
+                    LOG,
+                    "Block proposal {} with {} transactions, ready in {}ms",
+                    bestBlock::toLogString,
+                    bestBlock.getBody().getTransactions()::size,
+                    () -> System.currentTimeMillis() - startedAt);
               } else {
-                final var resultBest = validateBlock(bestBlock);
-                if (resultBest.blockProcessingOutputs.isPresent()) {
-                  mergeContext.putPayloadById(payloadIdentifier, bestBlock);
-                } else {
-                  LOG.warn(
-                      "failed to execute block proposal {}, reason {}",
-                      bestBlock.getHash(),
-                      resultBest.errorMessage);
-                }
+                LOG.warn(
+                    "Failed to execute block proposal {}, reason {}",
+                    bestBlock.getHash(),
+                    resultBest.errorMessage);
               }
             });
+  }
 
-    return payloadIdentifier;
+  private CompletableFuture<Block> retryBlockCreation(
+      final PayloadIdentifier payloadIdentifier,
+      final Supplier<Block> blockCreator,
+      final long remainingTime) {
+    final long startedAt = System.currentTimeMillis();
+
+    debugLambda(
+        LOG,
+        "Block creation started for payload id {}, remaining time is {}ms",
+        payloadIdentifier::toShortHexString,
+        () -> remainingTime);
+
+    return exceptionallyCompose(
+        ethContext
+            .getScheduler()
+            .scheduleComputationTask(blockCreator)
```

Comment:
Using EthScheduler provides much more flexibility, since it allow for some executor configuration, also metrics are provided out of the box, and testing is also supported.
The computation executor is not used during the normal lifecycle, it is used by some subcommands, and I am tuning it to better suits block proposal

Reply:
Sounds good! What new metrics will be available for block building?

Reply:
the metrics are related to the executors thread pool and task queue
![image](https://user-images.githubusercontent.com/91944855/191039037-e580407f-7000-4a30-abf1-99ccd380988d.png)
for block building we need to add them